File: wrap/android/src/main/kotlin/com/pistadb/PistaDBExtensions.kt
/*
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 */
@file:JvmName("PistaDBKt")

package com.pistadb

/**
 * Kotlin-idiomatic extensions and DSL builder for PistaDB.
 *
 * Usage:
 * ```kotlin
 * val db = pistaDB(filesDir.path + "/embeddings.pst", dim = 384) {
 *     metric    = Metric.COSINE
 *     indexType = IndexType.HNSW
 *     params { hnswEfSearch = 100 }
 * }
 *
 * db += VectorEntry(myEmbedding, label = "doc-42")
 * val hits = db.search(queryVec, k = 5)
 * hits.forEach { println("${it.label}  d=${it.distance}") }
 * ```
 */

/* ── DSL builder ─────────────────────────────────────────────────────────── */

/** Mutable configuration for the [pistaDB] DSL. */
class PistaDBConfig(val path: String, val dim: Int) {
    var metric: Metric       = Metric.L2
    var indexType: IndexType = IndexType.HNSW
    val params: PistaDBParams = PistaDBParams.defaults()

    /** Inline params configuration block. */
    fun params(block: PistaDBParams.() -> Unit) {
        params.block()
    }
}

/**
 * Opens or creates a PistaDB database using a configuration DSL.
 *
 * ```kotlin
 * val db = pistaDB(path, dim = 768) {
 *     metric    = Metric.COSINE
 *     indexType = IndexType.HNSW
 *     params    { hnswEfSearch = 50 }
 * }
 * ```
 */
fun pistaDB(path: String, dim: Int, block: PistaDBConfig.() -> Unit = {}): PistaDB {
    val cfg = PistaDBConfig(path, dim).apply(block)
    return PistaDB(cfg.path, cfg.dim, cfg.metric, cfg.indexType, cfg.params)
}

/* ── Operator overloads ──────────────────────────────────────────────────── */

/**
 * Inserts a vector using [VectorEntry] via the `+=` operator.
 *
 * ```kotlin
 * db += VectorEntry(floatArrayOf(...), "my-doc")
 * ```
 * @param entry Must carry a non-null [VectorEntry.vector]; requires that the
 *              caller supplies a unique id separately.
 */
operator fun PistaDB.plusAssign(pair: Pair<Long, VectorEntry>) {
    insert(pair.first, pair.second.vector, pair.second.label)
}

/** Retrieves a stored entry by id using the `[]` operator. */
operator fun PistaDB.get(id: Long): VectorEntry = get(id)

/* ── Batch helpers ───────────────────────────────────────────────────────── */

/**
 * Inserts an [Iterable] of (id, vector, label) triples.
 *
 * ```kotlin
 * db.insertAll(embeddings.mapIndexed { i, v -> Triple(i.toLong(), v, "doc-$i") })
 * ```
 */
fun PistaDB.insertAll(entries: Iterable<Triple<Long, FloatArray, String>>) {
    for ((id, vec, label) in entries) {
        insert(id, vec, label)
    }
}

/**
 * Returns search results as a [List] instead of an array.
 *
 * ```kotlin
 * val results: List<SearchResult> = db.searchList(queryVec, k = 10)
 * ```
 */
fun PistaDB.searchList(query: FloatArray, k: Int = 10): List<SearchResult> =
    search(query, k).toList()

/**
 * Returns only the ids from a search, sorted by ascending distance.
 *
 * ```kotlin
 * val nearestIds: List<Long> = db.searchIds(queryVec, k = 5)
 * ```
 */
fun PistaDB.searchIds(query: FloatArray, k: Int = 10): List<Long> =
    search(query, k).map { it.id }

/* ── use extension ───────────────────────────────────────────────────────── */

/**
 * Executes [block] with this database, saves, then closes – even if an
 * exception is thrown.
 *
 * ```kotlin
 * pistaDB(path, 384).use { db ->
 *     db.insert(1L, vec, "hello")
 * }  // saved and closed automatically
 * ```
 */
inline fun <T> PistaDB.use(block: (PistaDB) -> T): T {
    return try {
        block(this)
    } finally {
        close()
    }
}

/* ── PistaDBParams extensions ────────────────────────────────────────────── */

/** Fluent HNSW preset for high-recall scenarios. */
fun PistaDBParams.highRecallHnsw(): PistaDBParams = apply {
    hnswM              = 32
    hnswEfConstruction = 400
    hnswEfSearch       = 200
}

/** Fluent HNSW preset for low-latency scenarios. */
fun PistaDBParams.fastHnsw(): PistaDBParams = apply {
    hnswM              = 16
    hnswEfConstruction = 100
    hnswEfSearch       = 20
}

File: wrap/ios/Sources/PistaDB/PistaDBExtensions.swift
/*
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 */
import Foundation

// MARK: - DSL factory

/// Opens or creates a PistaDB database using a trailing-closure DSL.
///
/// ```swift
/// let db = try pistaDB(path: dbPath, dim: 384) {
///     $0.metric    = .cosine
///     $0.indexType = .hnsw
///     $0.params.hnswEfSearch = 100
/// }
/// ```
public func pistaDB(
    path: String,
    dim: Int,
    configure: (inout PistaDBConfig) -> Void = { _ in }
) throws -> PistaDB {
    var cfg = PistaDBConfig(dim: dim)
    configure(&cfg)
    return try PistaDB(path: path, dim: dim,
                       metric: cfg.metric,
                       indexType: cfg.indexType,
                       params: cfg.params)
}

/// Mutable configuration value used in the ``pistaDB(path:dim:configure:)`` DSL.
public struct PistaDBConfig {
    public var metric:    Metric        = .l2
    public var indexType: IndexType     = .hnsw
    public var params:    PistaDBParams = PistaDBParams()

    public init(dim: Int) {}

    /// Inline params shorthand: `cfg.params { $0.hnswEfSearch = 100 }`.
    public mutating func params(_ configure: (inout PistaDBParams) -> Void) {
        configure(&params)
    }
}

// MARK: - withDatabase (scoped use)

/// Opens a database, executes `body`, saves, then closes – even if `body` throws.
///
/// ```swift
/// try withDatabase(path: dbPath, dim: 384) { db in
///     try db.insert(id: 1, vector: vec, label: "hello")
/// }  // saved & closed automatically
/// ```
public func withDatabase<T>(
    path: String,
    dim: Int,
    metric: Metric = .l2,
    indexType: IndexType = .hnsw,
    params: PistaDBParams? = nil,
    body: (PistaDB) throws -> T
) throws -> T {
    let db = try PistaDB(path: path, dim: dim,
                         metric: metric, indexType: indexType, params: params)
    defer { db.close() }
    return try body(db)
}

// MARK: - PistaDB subscript

public extension PistaDB {
    /// Retrieves a stored entry by id using subscript syntax.
    ///
    /// ```swift
    /// let entry = try db[42]
    /// ```
    subscript(id: UInt64) -> VectorEntry {
        get throws { try get(id: id) }
    }
}

// MARK: - Batch from Dictionary

public extension PistaDB {
    /// Inserts a dictionary of `[id: (vector, label)]`.
    func insertBatch(_ dict: [UInt64: (vector: [Float], label: String)]) throws {
        for (id, pair) in dict {
            try insert(id: id, vector: pair.vector, label: pair.label)
        }
    }

    /// Inserts a dictionary of `[id: vector]` with empty labels.
    func insertBatch(_ dict: [UInt64: [Float]]) throws {
        for (id, vec) in dict {
            try insert(id: id, vector: vec)
        }
    }
}

// MARK: - PistaDBParams presets

public extension PistaDBParams {
    /// HNSW preset for high-recall scenarios.
    static var highRecall: PistaDBParams {
        var p = PistaDBParams()
        p.hnswM              = 32
        p.hnswEfConstruction = 400
        p.hnswEfSearch       = 200
        return p
    }

    /// HNSW preset for low-latency / on-device scenarios.
    static var lowLatency: PistaDBParams {
        var p = PistaDBParams()
        p.hnswM              = 16
        p.hnswEfConstruction = 100
        p.hnswEfSearch       = 20
        return p
    }
}

// MARK: - SearchResult helpers

public extension Array where Element == SearchResult {
    /// Returns only the ids in order.
    var ids: [UInt64] { map(\.id) }

    /// Returns only the distances in order.
    var distances: [Float] { map(\.distance) }

    /// Returns only the labels in order.
    var labels: [String] { map(\.label) }

    /// Returns the closest result (first element).
    var nearest: SearchResult? { first }
}
